docs/source/libraries.rst

```diff
@@ -4,79 +4,197 @@
+
+   setup.py
+   my_great_package/
+      __init__.py
+      stuff.py
+      py.typed
+
+It's important to ensure that the ``py.typed`` marker file is included in the
+distributed package. If using ``setuptools``, this can be achieved like so:
+
+.. code-block:: python
+
+   from setuptools import setup
+
+   setup(
+      name="my_great_distribution",
+      version="0.1",
+      package_data={"my_great_package": ["py.typed"]},
+      packages=["my_great_package"],
+   )
+
+
+Type stub files included in the package
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+It's possible to include a mix of type stub files (``.pyi``) and inline type
+annotations (``.py``). One use case for including type stub files in your
+package is to provide types for extension modules in your library. A typical
+directory structure would look like:
+
+.. code-block:: text
+
+   setup.py
+   my_great_package/
+      __init__.py
+      stuff.py
+      stuff.pyi
+      py.typed
+
+If using ``setuptools``, we can ensure the ``.pyi`` and ``py.typed`` files are
+included like so:
+
+.. code-block:: python
+
+   from setuptools import setup
+
+   setup(
+      name="my_great_distribution",
+      version="0.1",
+      package_data={"my_great_package": ["py.typed", "stuff.pyi"]},
+      packages=["my_great_package"],
+   )
 
-In many existing type stubs (such as those found in typeshed), default
-parameter values are replaced with with ``...`` and all docstrings are
-removed. We recommend that default values and docstrings remain within
-the type stub file so language servers can display this information to
-developers.
+The presence of ``.pyi`` files does not affect the Python interpreter at runtime
+in any way. However, static type checkers will only look at ``.pyi`` file and
+ignore the corresponding ``.py`` file.
 
-Library Interface
-=================
+Companion type stub package
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-`PEP 561 <https://www.python.org/dev/peps/pep-0561/>`__ indicates that a
-``py.typed`` marker file must be included in the package if the author
-wishes to support type checking of their code.
+These are often referred to as "stub-only" packages. The package should have a
+prefix of the runtime package name with a suffix of ``-stubs``. The py.typed
+marker file is not necessary for stub-only packages. This approach can be useful
+to develop type stubs independently from your library.
+
+For example:
+
+.. code-block:: text
+
+   setup.py
+   my_great_package-stubs/
+      __init__.pyi
+      stuff.pyi
+
+
+.. code-block:: python
+
+   from setuptools import setup
+
+   setup(
+      name="my_great_package-stubs",
+      version="0.1",
+      package_data={"my_great_package-stubs": ["__init__.pyi", "stuff.pyi"]},
+      packages=["my_great_package-stubs"]
+   )
+
+
+Users are then able to install the stubs-only package separately to provide
+types for the original library.
+
+Note that to ensure inclusion of ``.pyi`` and ``py.typed`` files in an sdist
```

Comment:
Weird that this is in the stub package section, while the one about inline stub files only talks about `setup.py`.